AppController/djinn.rb

```diff
@@ -5357,25 +5301,95 @@ def find_lowest_free_port(starting_port, ending_port=0, appid="")
-  # that each application has received as well as the number of requests that
-  # are sitting in haproxy's queue, waiting to be served.
-  def scale_appservers_within_nodes
+  # Scale AppServers up/down for each application depending on the current
+  # queued requests and load of the application.
+  #
+  # Returns:
+  #   An Integer indicating the number of AppServers that we couldn't
+  #   start for lack of resources.
+  def scale_appservers
+    needed_appservers = 0
     @apps_loaded.each { |app_name|
       initialize_scaling_info_for_app(app_name)
 
       # Get the desired changes in the number of AppServers.
       delta_appservers = get_scaling_info_for_app(app_name)
       if delta_appservers > 0
         Djinn.log_debug("Considering scaling up app #{app_name}.")
-        try_to_scale_up(app_name, delta_appservers)
+        unless try_to_scale_up(app_name, delta_appservers)
+          needed_appservers += delta_appservers
+        end
       elsif delta_appservers < 0
         Djinn.log_debug("Considering scaling down app #{app_name}.")
         try_to_scale_down(app_name, delta_appservers.abs)
       else
         Djinn.log_debug("Not scaling app #{app_name} up or down right now.")
       end
     }
+
+    return needed_appservers
+  end
+
+
+  # Adds or removes AppServers and/or nodes to the deployment, depending
+  # on the statistics of the application and the loads of the various
+  # services.
+  def scale_deployment
+    needed_appservers = scale_appservers
+    if needed_appservers > 0
+      Djinn.log_debug("Need to start VMs for #{needed_appservers} more AppServers.")
+    end
+
+    # Here we cound the number of machines we need to spawn, and the roles
+    # we need.
+    vms_to_spawn = 0
+    roles_needed = {}
+    if needed_appservers > 0
+      Integer(needed_appservers/3).downto(0) {
+        vms_to_spawn += 1
+        roles_needed["appengine"] = [] unless roles_needed["appengine"]
+        roles_needed["appengine"] << "node-#{vms_to_spawn}"
+      }
+    end
+
+    # Check if we need to spawn VMs and the InfrastructureManager is
+    # available to do so.
+    return unless vms_to_spawn > 0
+    if SCALE_LOCK.locked?
+      Djinn.log_debug("Another thread is already working with the InfrastructureManager.")
+      return
+    end
+
+    Thread.new {
+      SCALE_LOCK.synchronize {
+        Djinn.log_info("We need #{vms_to_spawn} more VMs.")
+
+        if Time.now.to_i - @last_scaling_time < (SCALEUP_THRESHOLD *
+              SCALE_TIME_MULTIPLIER * DUTY_CYCLE)
+          Djinn.log_info("Not scaling up right now, as we recently scaled " +
+            "up or down.")
+          return
+        end
+
+        @state_change_lock.synchronize {
+          allowed_vms = Integer(@options['max_images']) - @nodes.length
+          if allowed_vms < vms_to_spawn
+            Djinn.log_warn("Requested #{vms_to_spawn} VMs, but we can only" +
+              " start #{allowed_vms}, so not spawning more nodes.")
```

Comment:
So I was testing autoscaling with a 2 node deployment and I set max images to 3 and appengine to 15. It needed to spawn 5 VMs to accommodate 15 appservers, that makes vms_to_spawn = 5 and allowed_vms = 1 (according to this calculation). Shouldn't it still spawn that one VM which is still in the max_images limit? It would not do it in this condition because allowed_vms is less than the vms_to_spawn.